mlir/lib/IR/AffineExpr.cpp

```diff
@@ -406,12 +419,25 @@ static bool isDivisibleBySymbol(AffineExpr expr, unsigned symbolPos,
     AffineBinaryOpExpr binaryExpr = cast<AffineBinaryOpExpr>(expr);
     if (opKind != expr.getKind())
       return false;
-    return isDivisibleBySymbol(binaryExpr.getLHS(), symbolPos, expr.getKind());
+    if (llvm::any_of(visitedExprs, [](auto expr) {
+          return expr.getKind() == AffineExprKind::Mul;
+        }))
+      return false;
+    return isDivisibleBySymbolImpl(binaryExpr.getLHS(), symbolPos,
+                                   expr.getKind(), visitedExprs, depth + 1);
```

Comment:
Assuming the logic here is correct, you can replace the `visitedExprs` vector with just a boolean flag that tells you whether you're in a mul.

Reply:
flag can't be used, If this expression`(s0 * 2) + s0 ceildiv 2)`,I need to reset the flag after visited lhs `add`,Do you think this behavior is easy to understand?

Reply:
Maybe I'm missing something. Doesn't your call tree looks like this:

```             
              flag
+             false
   *          false
      s0      true
      2       true
   ceildiv    false
      s0      false
      2       false
```

Reply:
Yes.we need to reset flag to false,before visit ceildiv.

Reply:
But these are sibling branches of the recursion. The flag is an argument, so it'll be false in ceildiv and below because there's no mul in the call chain.

Reply:
Ok.I see.

Reply:
@jreiffers Done.